cmds/dhclient/dhclient.go

```diff
@@ -46,17 +47,27 @@ var (
 )
 
 func ifup(ifname string) (netlink.Link, error) {
-	iface, err := netlink.LinkByName(ifname)
-	if err != nil {
-		return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
-	}
-
-	if err := netlink.LinkSetUp(iface); err != nil {
-		return nil, fmt.Errorf("%v: %v can't make it up: %v", ifname, iface, err)
-	}
+	start := time.Now().Second()
+	for time.Now().Second() - start < linkUpAttempt {
+		// Note that it may seem odd to keep trying the
+		// LinkByName operation, by consider that a hotplug
+		// device such as USB ethernet can just vanish.
+		iface, err := netlink.LinkByName(ifname)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get interface by name %v: %v", ifname, err)
+		}
 
-	return iface, nil
+		if iface.Attrs().OperState == netlink.OperUp {
+			debug("Link %v is up", ifname)
+			return iface, nil
+		}
 
+		if err := netlink.LinkSetUp(iface); err != nil {
+			return nil, fmt.Errorf("%v: %v can't make it up: %v", ifname, iface, err)
+		}
+		time.Sleep(1 * time.Second)
+	}
+	return nil, fmt.Errorf("Link %v still down after 30 seconds", ifname)
```

Comment:
The "30" should probably be a format option for "linkUpAttempt". That way, it will be easier to change in the future.